### DISCUSSION TITLE
0602 피어세션일지

### DISCUSSION THREAD

**Discussion by gvsteve24** [bcaitech1/p4-dkt-freshtomato#66] [General]
# 오늘의 이모지 + TMI

|    아라     |      진현       |    민용    |  재훈   |    정현    |   유라   |
| :---------: | :-------------: | :--------: | :-----: | :--------: | :------: |
| :milky_way: | :space_invader: | :question: | :peach: | :pig_nose: | :butter: |

# 토론 및 회의

ara_feature 뺐을 때 0.807 -> 0.0804 로 떨어짐

* shift 할때 시간이 많이 걸린다.

train. valid  나누는 기준을 다시 정해서 쓰고 있다. 

* 관련해서 인사이트를 얻으면 좋겠다.

* 관련 설명 스크린샷

![image-20210602163700439](https://user-images.githubusercontent.com/28102768/120468687-764e5a80-c3dc-11eb-894f-9ad137464b1e.png)

> 기존에 train 데이터 셋을 한번만 잘라서 쓰는 것과 K-fold의 차이점 
> => 한번만 잘라서 쓰면 validset에 대한 오버피팅을 발생시킨다.

![image-20210602164112436](https://user-images.githubusercontent.com/28102768/120468707-7cdcd200-c3dc-11eb-9318-73ca66ca241e.png)

> AutoML의 경우 Pycaret을 쓰고 있는데 Issue에서 test와 train을 다르게 설정하는 파라메터가 없다고 함. setup() 함수의 파라메터에서 오직 비율만 나온다고 함. 
>
> =>  결국 holdout으로 하나 빼주는 걸 커스터마이징하려고 함
> => pycaret-nightly가 있긴 있음. 근데 아직 익히지 못했으므로 커스터마이징해서 시도

![image-20210602164516252](https://user-images.githubusercontent.com/28102768/120468736-8403e000-c3dc-11eb-812d-e9a765ecc1b5.png)

> hold-out rate : 처음에 unseen 데이터로 빼주려는 비율
>
> 이제 위와 같이 holdout 을 userID기준으로 잡는데 이때 test의 경향과 비슷하기 위해서 마지막에 푼 문제와 같은 경우로 만들게 됨.

![image-20210602170647331](https://user-images.githubusercontent.com/28102768/120468754-87976700-c3dc-11eb-834f-c5a36cf02be3.png)

> test문제있는 애들은 꽤 될거고 그중에 얼마나 뽑아야되는지 혹은 마지막에 있는 애들을 뽑아야되는지 어렵겠다.

DL의 경우는 어떻게 하는 것이 좋을지?

>  train에서 마지막 문제 로우만 빼는 것을 hold out이라고 하면 어떨까?

* data leakage가 있는 것 아니냐 (피처나 이런것들을 이미 썼으므로)
* 대안) holdout을 빼는 것은 사용자별로 빼면 어떨까 근데 그러면 너무 제외되는게 많을 것 같다. 

#### AutoML _ pycaret

앙상블을 하면 최종 모델까지 설정해준다.

ML을 위한 모델이라서 DL에서는 Sweep을 쓰는게 낫겠다.

sweep을 할때 beysian으로 경계값을 정해서 쓰는건지?

* sweep은 어떤 py파일을 선택해야되서 for문 안에 넣기 어렵다. fold로는 어떻게 해야되는지 모르겠다.

* optuna는 가능하긴 함.

* config작성하고 score(예를들면)를 optimizing한다.

* kfold안에 아래의 코드를 작성해주면 됩니다. 예시)

  ```python
  study = optuna.create_study('maximize')
  study.optimze(objectiv, n_trials=100)
  ```

  

### DKT 논문 리뷰 _ 민용

과거에 사용되었던 Markov모델의 한계를 RNN으로 극복하였고 실제로 25%의 성능향상을 가져왔다. 

Prerequisite : Vanilla RNN, LSTM, GRU

* 이전에 문제들을 기준으로 다음 문제를 맞출지 못맞출지를 확인하기 때문에 문제들간의 관계를 알 수 있다. 
* 각 문제 단계별 Knowledge state를 바탕으로 어떤 문제를 풀었을 때 Knowledge state가 최적화되는지 계산할 수 있다. 따라서 최적의 커리큘럼을 디자인할 수 있다.

# 저녁 메뉴 !

| 재훈   | 아라   | 민용       | 유라       | 진현       | 정현   |
| ------ | ------ | ---------- | ---------- | ---------- | ------ |
| 갈비탕 | 파스타 | :question: | :question: |샐러드 | 갈비탕 |

**Comment by arabae** [comment]
일지 감사합니당 :)
  
---
### TODO
- [x]  [실험] notebook에 올려주신 SAINT 모델 baseline 실험 후 성능 확인
- [x]  [실험] AraFeature만 추가하여 SAINT 모델 훈련 후 성능 비교
- [x]  [실험] 기존의 AraFeature 중 <s>정답, 오답개수와 완련된 feature를 continuous로 변경</s>하여 실험
    - [x]  우선 baseline에서 제공해준 feature와 정답, 오답 개수 feature 제거하여 실험 진행 (+ problem_number를 추가) 
    - [x]  이를 SAINT에도 적용
- [x]  [네트워킹] 이력서 포맷 정하기
- [ ]  [네트워킹] 이전에 채워두었던 내용까지는 변경된 이력서에 옮겨두기
- [x] [논문] RELATED WORKS, PROPOSED MODEL 읽기 -- PROPOSED MODEL 중 Deep Self-Attentive Encoder-Decoder까지

**Comment by gvsteve24** [comment]
- [x] [실험] 오피스아워: process_batch, LSTM, LSTMATT, BERT 중간결과물 직접 확인하기
- [ ] [강의] Transformer Tuning 듣기
- [ ] [취업] 이력서 프로젝트 소개 항목 작성하기
- [ ] [논문] DKVMN Related works 까지 ppt에 정리하기

**Comment by KimJinHye0n** [comment]
![슬라이드1](https://user-images.githubusercontent.com/63627253/120486724-46a94d80-c3f0-11eb-8ece-485545de70ac.JPG)
![2](https://user-images.githubusercontent.com/63627253/120491140-f502c200-c3f3-11eb-87b2-11ddb1d5b350.jpg)
![슬라이드3](https://user-images.githubusercontent.com/63627253/120486743-4a3cd480-c3f0-11eb-819f-c44736e0bca4.JPG)

저걸로 정리가 될지는 모르겠지만 개인적으로 정리해봤습니다.

#### TODO
1. Transformer 모델 돌려보기
2. Saint 논문 읽기

**Reply by arabae** [reply]
헐..ㅠㅠㅠ 뭐야 너무 완벽하자나 edit 이용해서 본문 당장 이걸로 추가해주세여!! 너무 좋습니다

**Reply by MignonDeveloper** [reply]
깔-끔 그자체.... 멋있어,,,

**Reply by MoonJaeHoon** [reply]
진짜 미쳤다.... 내가 발표자료를 이렇게 만들어야 하는데.. 개깔끔하다

**Reply by KimJinHye0n** [reply]
holdout 완벽하게 정해지면 그때 Discussion에 올릴께 발표자료는 맡겨주시죠

**Reply by gvsteve24** [reply]
와 정말 금손이야.... 그리구 너무 고마워 😢 정리를 완벽하게 만들어줌

**Comment by MoonJaeHoon** [comment]
오늘할일

- [x]   최고점수였던 0.8072 실험에서 => 내 column들 중 결측치 있는 column들 모두 채워주고 모델링하여 점수 상승하는지 확인하기

=> 단순히 0으로 채워줬더니 떨어짐.. ( 0.8072 -> 0.7957 )

- [x]  위 실험이 잘 되는지 확인 후 0.8048 실험에 추가적용해보기

    두번째 점수였던 0.8048 실험 (민용이와 아라꺼 추가)으로부터 더 나아가서 실험해보기
    

- [x]     2-1) 1번이 잘된다면 ->

     => ~모든 컬럼 결측치 모두 채워주고 모델링하여 점수 보기~  ( 0.8048 -> 0.8017 )

- [x]     2-2) 1번이 잘 안된다면

    => 아라 feature들에만 결측치 모두 채워주고 모델링하여 점수 보기

- [ ]  적절한 HoldoutSet & TrainSet 찾기... (실험 계속 해야하는 부분) ( 0.8048 -> 0.8019 )

- 아이디어1 : 이전까지 다양하게 나눠본 결과 train과 Holdout의 정답률 분포가 매우 차이가 많이 났었다..
  - 따라서 train의 정답률 분포가 최대한 HoldoutSet에 가까워지게끔 만들어서 train해보는건 어떨까?
  - User 별 맨 마지막 row만 쓰면 분포가 비슷한듯 (하지만 이러면 학습가능한 데이터 수가 너무 적다)

**Reply by Yuuraa** [reply]
- [x]  LightGBM 파라미터 공부하기
- [ ] Pycaret으로 하이퍼파라미터 튜닝
- [ ] Hold-out + K-Fold 구현하기
